backend/apps/monkvision/apis/lib/utils.js

```diff
@@ -28,4 +30,48 @@ exports.fromSQLiteToUTCOrLocalTime = function(sqliteDate, notUTC) {
   */
 exports.getTimeRangeForSQLite = (range) => {
     return {from: exports.fromISOToSQLite(range.from), to: exports.fromISOToSQLite(range.to)}
-}
\ No newline at end of file
+}
+
+/**
+ * Fetches data from the database and caches it if necessary.
+ * @param {string} queryFunc - The database query function name.
+ * @param {string} id -  Log ID Unique identifier.
+ * @param {Object} timeRange - Time range (optional).
+ * @param {...any} extraParams - Additional parameters to pass to the query function.
+ * @returns  Returns an array of fetched or cached data.
+ */
+const cache = {}; 
+
+async function fetchAndCache(id, timeRange, queryFunc, ...extraParams) {
+    const rows = await db[queryFunc](id, timeRange, ...extraParams);
+    cache[id].data.push(...rows);
+    return rows;
+}
+
+async function fetchIncrementalAndCache(id, range, queryFunc, prepend = false, ...extraParams) {
+    const rows = await db[queryFunc](id, range, ...extraParams);
+    if (prepend) cache[id].data.unshift(...rows);
+    else cache[id].data.push(...rows);
+    return rows;
+}
+
+exports.fetchDBData = async function (queryFunc, id, timeRange, ...extraParams) {
+    if (!timeRange) return db[queryFunc](id, ...extraParams);
+    if (!cache[id]) cache[id] = { data: [], lastRange: null };
+    const { lastRange } = cache[id];
+
+    if (!lastRange || (timeRange.from < lastRange.from && timeRange.to > lastRange.to) ) {
+        cache[id].data = await fetchAndCache(id, timeRange, queryFunc, ...extraParams);
+        cache[id].lastRange = { ...timeRange };
+    } else if (timeRange.to > lastRange.to && timeRange.from === lastRange.from) {
```

Comment:
Also add proper parenthesis in if conditions!!
WRONG -> `if (timeRange.to > lastRange.to && timeRange.from === lastRange.from)`
RIGHT -> `if ((timeRange.to > lastRange.to) && (timeRange.from === lastRange.from))`